rfcs/text/0428-extending-and-duplicating-types.md

```diff
@@ -0,0 +1,970 @@
+  "$id": "https://example.com/schema",
+
+  // ... schema contents  ...
+
+  "$defs": {
+    "closed": {
+      "$anchor": "closed",
+      "$ref": "#",
+      "unevaluatedProperties": false
+    }
+  }
+}
+```
+
+Here, referencing `https://example.com/schema` in a `$ref` will result in an _open schema_ that _does not_ specify `{ "unevaluatedProperties": false }`. Referencing `https://example.com/schema#closed` results in retrieving a _closed schema_ that _does_ specify `{ "unevaluatedProperties": false }`.
+
+**Closed by default schema**:
+
+```json
+{
+  "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
+  "$id": "https://example.com/schema",
+  "$ref": "#open",
+  "unevaluatedProperties": false,
+
+  "$defs": {
+    "schema": {
+      "$anchor": "open"
+      // ... schema contents ...
+    }
+  }
+}
+```
+
+Here, referencing `https://example.com/schema` in a `$ref` will result in a _closed schema_ that _does_ specify `{ "unevaluatedProperties": false }`. Referencing `https://example.com/schema#open` results in retrieving an _open schema_ that _does not_ specify `{ "unevaluatedProperties": false }`.
+
+(Thanks Jason Desrosiers for the suggestions!)
+
+Using the above setup would mean that we need to specify `#open` at the end of type URLs when URLs appear in `allOf`. We would also need to serve type schemas with one of the above structures, such that they conform with JSON Schema. Although this does add some ceremony around extending types, we would end up with a type extension system that would conform to JSON Schema without implicitness or redefining keyword semantics.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+- [JSON Schema composition](https://json-schema.org/understanding-json-schema/reference/combining.html) and [`unevaluatedProperties`](https://json-schema.org/understanding-json-schema/reference/object.html#unevaluated-properties)
+- [Programming languages subtyping](https://en.wikipedia.org/wiki/Subtyping)
+- [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- We haven't specified how projecting/selecting properties of a supertype from a subtype instance is possible. It is an open question how we actually pick out the exact properties of a subtype to provide a valid supertype instance in embedding applications.
+- The current argument for not allowing cyclic type hierarchies mostly build on a feeling that type hierarchies shouldn't be too indirect/obfuscated, but there could be stronger arguments for allowing/disallowing it.
+- The decision to have type substitution between subtype and supertype does mean the extended types are restrictive in that they do not allow for removing/changing existing properties. Without a way to transform data between types (mappings), allowing overrides is not possible while keeping compatibility.
```

Comment:
I think we mention mappings too many times in this RFC without explaining what "mappings" are. 

They're currently an internal only thought, and we don't have a good resource to point to, so this would be a good opportunity to raise the general idea of them, and we can look at publishing a draft placeholder RFC to create something referable soon.

Reply:
Happy to help with this.